### PR TITLE
Support multiple python using tox with Poetry

### LINKED ISSUE #73: Support latest Python version
Hi! Thank you for developing such a great tool!
Currently, Hera supports only Python 3.7. The version is now security status, will become end of support soon.

reference: https://www.python.org/downloads/

So it is better that supporting Python 3.8+, IMO.
I'm happy to create PR for supporting newer version of Python, if it is OK.

Thank you.

### PR DESCRIPTION
**Pull request by hirosassa**
fix #73 

## What is this PR?

Related to #93, I implemented support for multiple Python versions using `tox`.
Different from #93, this PR uses `Poetry` as a package management to simplify configurations of linters, tests and builds.

### CODE REVIEW

**Comment by hirosassa** [comment]
[note] 

I said

> applying all of the solutions is not preferable for the developers and such a big change will cause bugs

but @harelwa implemented nice PR to fix same issue using pipenv.
So I implemented another version to compare.


@harelwa @bobh66 @flaviuvadan plz review!

**Comment by harelwa** [comment]
Hi @hirosassa 

That's great !!

I think we [ as in `hera` project ] needs to decide on the following two things, and unite the focus into a single PR ( this + #93 + #92 ).

1. The `makefile`

In your changes here, you have basically replaced the `make` + `Makefile` with `tox` + `tox.ini`

I'd love if we can address what I wrote in the `Makefile` section of #93 description, which is also related to my answer to you here regarding `setup.cfg` - https://github.com/argoproj-labs/hera-workflows/pull/93#discussion_r805132881

2. Drop `Pipenv` In favour of `Poetry`

@flaviuvadan @bobh66 @bchalk101 @GuyPozner ?

**Comment by flaviuvadan** [comment]
Please see my comment [here](https://github.com/argoproj-labs/hera-workflows/pull/93#issuecomment-1038459672). I think Hera should introduce the usage of `tox` at first to get a stable build, release 1-2 features using that, so we test that setup. The introduction of `tox` gives us the benefit of testing the package in multiple Python versions, which is an amazing start for the goal of supporting multiple Python versions! After 1-2 releases, I suggest we introduce this change 🙂 I am excited about the ideas you folks brought for improving Hera's testing and stability, so I am excited to get this rolling as well.

File: tox.ini
Comment (flaviuvadan) [inline tox.ini:1]: So now, rather than invoking `make test`, we invoke `tox -e test`, right? If so, this is great! What do you think about keeping the `Makefile` around but replacing the inner calls with `tox -e`? 🙂 Curious whether that might be nicer for people who may know `make` but not necessarily `tox`.

**Comment by flaviuvadan** [comment]
@hirosassa I think we can merge this soon! Tested the CICD improvements @harelwa and everything deployed just fine 🚀

**Comment by hirosassa** [comment]
@flaviuvadan Thanks for informing this!

File: tox.ini
Comment (hirosassa) [reply tox.ini:1]: > So now, rather than invoking make test, we invoke tox -e test, right?

There's no exact counterpart of `make test` in this branch.
`tox` is the task runner on each isolated environment.
If your want to run test for Python 3.8, then you should invoke `tox -e py38`. 
`make lint`, `make format` correspond to `tox -e lint` `tox -e format`

How about making Makefile below?
```Makefile
testall: # run all test in all Python version
	tox

test-pyXX: # run test on PythonXX
	tox -e pyXX
...
```

**Comment by flaviuvadan** [comment]
@hirosassa now that we merged the CICD changes, I think we can pick this up again and deliver the migration to Poetry!

File: .github/workflows/cicd.yaml
Comment (hirosassa) [inline .github/workflows/cicd.yaml:123]: For poetry, `publish` sub-command builds package and uploads to PyPI with `py3-none-any` wheel.

File: .github/workflows/cicd.yaml
Comment (hirosassa) [inline .github/workflows/cicd.yaml:74]: This jobs is removed because `poetry` publish this automatically

File: Makefile
Comment (hirosassa) [inline Makefile:14]: I integrate `isort` task with `format` task (isort is also format).

File: pyproject.toml
Comment (hirosassa) [inline pyproject.toml:3]: I naively implemented `version`. I prefer using [poetry-dynamic-versioning](https://github.com/mtkennerly/poetry-dynamic-versioning) to add version by `git tag`.

@flaviuvadan Do you have any ideas on this?

**Review by hirosassa — COMMENTED** [review]
@flaviuvadan I updated. Please review!

File: .github/workflows/cicd.yaml
Comment (hirosassa) [reply .github/workflows/cicd.yaml:123]: For example, this [repo](https://github.com/m3dev/gokart) (I'm a mentainer of this) use poetry publish in [CI](https://github.com/m3dev/gokart/blob/master/.github/workflows/publish.yml#L27), then PyPI has [this files](https://pypi.org/project/gokart/#files)

File: pyproject.toml
Comment (flaviuvadan) [reply pyproject.toml:3]: I quite like the verbosity of `VERSION` thus far but I am definitely open to other approaches! Does the version need to be specified here? I wonder if we can leave the `setup.py` as it is?

File: .github/workflows/cicd.yaml
Comment (flaviuvadan) [inline .github/workflows/cicd.yaml:52]: Does tox run over the specified versions [here](https://github.com/argoproj-labs/hera-workflows/blob/main/tox.ini#L2)? Also, will it automatically run `testenv` in this case?

Comment (hirosassa) [reply .github/workflows/cicd.yaml:52]: But, this maybe confusing for the users who are not familiar with tox and github actions

Comment (hirosassa) [reply .github/workflows/cicd.yaml:52]: I agree with you! I'll fix this.

File: .github/workflows/cicd.yaml
Comment (hirosassa) [reply .github/workflows/cicd.yaml:16]: This verbosity comes from difference between [current main](https://github.com/argoproj-labs/hera-workflows/blob/main/tox.ini#L2) and [this branch](https://github.com/hirosassa/hera-workflows/blob/support-latest-python/tox.ini#L2).
The merit of this branch implementation is that if you'd like to run test, you type `tox -e py38` for example.
In the current main, you should type `tox -e python3.8`.
The frequency of changes of github actions configuration is low (once or twice in a year?), but running tests is frequent, so this verbosity is acceptable, IMO.

File: pyproject.toml
Comment (hirosassa) [reply pyproject.toml:3]: > I quite like the verbosity of VERSION

Yeah, I understand this 👍. This is very simple and easy to understand what the contributors should do. 
Currently we can't use VERSION file on pyproject.toml (pyproject.toml is `toml` so we could not load file dynamically).

> I think it's ok to leave this as is but update the contribution guideline to suggest that users increment this file rather than VERSION.

I updated this [here](https://github.com/argoproj-labs/hera-workflows/pull/94/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R15)

File: .github/workflows/cicd.yaml
Comment (flaviuvadan) [reply .github/workflows/cicd.yaml:16]: Got it. Thank you so much for the thoughtful response 🚀

File: .github/workflows/cicd.yaml
Comment (flaviuvadan) [inline .github/workflows/cicd.yaml:101]: Any chance we can uncomment this? I think it was added for information purposes and also in preparation for adding the coverage badge 🤔 so we know what files can be accessed.

File: README.md
Comment (flaviuvadan) [inline README.md:91]: Can we remove this section of the doc since the Makefile was removed in this PR?

File: tox.ini
Comment (flaviuvadan) [inline tox.ini:73]: What's the `#, format` comment? 🤔

**Review by flaviuvadan — COMMENTED** [review]
Left some very minor questions and tasks to trim some part of the documentation that still mentions the `make` command. This is so cool @hirosassa. I am very thankful for the contribution. I know this has been open for a while and I am so glad we're finally converging on the final version 🙂 I am going to merge this after some of the small cleanup and main merge into this so I will be on the lookout.

File: tox.ini
Comment (hirosassa) [reply tox.ini]: @flaviuvadan These lines are for `tox-gh-actions`,  so I removed these.

File: README.md
Comment (hirosassa) [reply README.md]: Thanks this is I missed. Removed.

File: .github/workflows/cicd.yaml
Comment (hirosassa) [reply .github/workflows/cicd.yaml]: Recovered!

**Comment by hirosassa** [comment]
@flaviuvadan Thank you for your comments! I applied the changes. And I found conflicting on `VERSION` .
It looks I can't resolve because I don't have write access on this repo.

**Review by flaviuvadan — APPROVED** [review]

**Comment by hirosassa** [comment]
@flaviuvadan Thank you for the fruitful discussions!

**Pull request merged**
